src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/CpsPackageReferenceProject.cs

```diff
@@ -371,6 +371,56 @@ public override Task<string> GetCacheFilePathAsync()
             return Task.FromResult(NoOpRestoreUtilities.GetProjectCacheFilePath(cacheRoot: spec.RestoreMetadata.OutputPath));
         }
 
+        // To avoid race condition, we work on copy of cache.
+        protected override (List<FrameworkInstalledPackages> installedPackagesCopy, List<FrameworkInstalledPackages> transitivePackagesCopy) GetCacheCopy()
+        {
+            return (new List<FrameworkInstalledPackages>(InstalledPackages), new List<FrameworkInstalledPackages>(TransitivePackages));
+        }
+
+        protected override void UpdatePackageListDetails(List<FrameworkInstalledPackages> packages, IEnumerable<FrameworkInstalledPackages> detectedNewPackages)
+        {
+            if (packages == null || detectedNewPackages == null)
+            {
+                return;
+            }
+
+            var dict = new Dictionary<NuGetFramework, FrameworkInstalledPackages>();
+
+            foreach (FrameworkInstalledPackages installedPackage in packages)
+            {
+                dict[installedPackage.TargetFramework] = installedPackage;
+            }
+
+            foreach (FrameworkInstalledPackages detectedNewInstalledPackage in detectedNewPackages)
+            {
+                if (detectedNewInstalledPackage == null)
```

Comment:
In theory, we should not get nulls, but this condition avoids any further problems. Good catch!